Implement functions necessary for TLS/SSL

facekapow opened the pull request:
This is currently a work-in-progress, but I'm opening up this PR early on to receive feedback on the code and make sure I'm implementing everything correctly. It's also easier to review the code incrementally rather than have to review huge chunks of code later on after I've completely implemented everything.

Things implemented so far:
- [x] DER encoding/decoding
- [ ] RSA cryptography
  - [x] RSA PKCS #1 v1.5 signature verification
- [x] CCN serialization/deserialization
- [x] EMSA PKCS #1 v1.5 encoding
- [x] Important CCN and CCZP operations
  - [x] `ccn_mul`
  - [x] `cczp_mod`
  - [x] `cczp_power`
- [ ] Elliptic Curve cryptography
  - [x] EC key generation
  - [x] EC shared secret computation
  - [x] EC signature verification
- [ ] AES cryptography
  - [x] AES block cipher
  - [x] AES-GCM
- Other stuff? I'm not entirely sure what is required for TLS/SSL, so I'm going to try implementing as many functions as I can

facekapow committed:
Implement a majority of DER encoding/decoding logic

facekapow committed:
Implement cc_print and ccdigest_oid_lookup

facekapow committed:
Fix OID decoding for DERs

facekapow committed:
Fix behavior of a few CCN operations and implement a couple more

Most notably, `ccn_mul` and `ccn_div_euclid` are now implemented (though only `ccn_mul` has been tested to be working).
In the fixes, the three biggest changes were:
  - `ccn_bitlen` now returns the number of *bits* by multiplying by CCN_UNIT_BITS (which is 64); it was previously multiply by CCN_UNIT_SIZE (which is 8)
  - `ccn_bitlen` checks to make sure the `last_used_unit` isn't 0, because `clz` performs undefined behavior in that case
  - `ccn_add` will now check for overflow against either input
Most of the other fixes were changes in counter types to match up with the types of their termination values (i.e. comparing cc_size with cc_size instead
of int with cc_size as it was previously).

facekapow committed:
Implement EMSA-PKCS #1 v1.5 encoding

facekapow committed:
Properly implement cczp_mod and cczp_power

facekapow committed:
Implement RSA PKCS #1 v1.5 signature verification

ahyattdev commented on src/ccrsa.c:
```diff
@@ -41,91 +43,150 @@ const uint8_t *ccder_decode_rsa_priv(const ccrsa_full_ctx_t key, const uint8_t *
-	cc_unit *s = alloca(sig_bytes);
-	if (ccn_read_uint(sig_units, s, sig_len, sig))
-	{
+	s = malloc(sig_bytes);
```
We are trying to avoid using malloc and instead use alloca so that it will work in the absence of a heap, like the original.

It's still used in some places but will be removed soon.

facekapow replied:
Got it. I was just confused at first by the mix of `alloca` and `malloc` in the previous code so I defaulted to `malloc`.

facekapow commented on src/ccder.c:
```diff
@@ -141,62 +302,161 @@ uint8_t *ccder_encode_eckey(size_t priv_size, const uint8_t *priv_key,
+	const uint8_t* oid_start = ccder_decode_tl(CCDER_OBJECT_IDENTIFIER, &oid_len, der, der_end);
+	size_t total_len = (size_t)(oid_start - der) + oid_len;
+
+	uint8_t* copy = malloc(sizeof(uint8_t) * total_len + 1);
```
Is the use of `malloc` here and in `ccder_decode_bitstring` still ok? We can't use `alloca` because we need to return the values.

This is assuming, of course, that the original Apple API was supposed to create copies and not just return pointers to the values in the DER structure. I couldn't really find any uses of the DER encoding/decoding API in open source software (Google only returned results for the copyrighted Apple CoreCrypto source, and we can't look at that for obvious reasons) and there's no header documentation either.

bugaevc replied:
(I'm not a lawyer, but) you can look at Apple's CoreCrypto, you just cannot, like, copy-paste from there (or write virtually the same code yourself). But you can look at it to figure out whether it mallocs returned memory or what.

ahyattdev replied:
Hmmm. Is it possible to use the memory between `der` and `der_end` for the return?

Usually what they do in this case is use some of the memory that was passed in.

ahyattdev replied:
(or at least some of the memory, then return a pointer to it's start)

LubosD replied:
We can do this in a clean-room way, if you want. You can ask questions and we give answers.

No memory is allocated in this function. The pointer it gets from ccder_decode_tl() is expected to be persistent.

facekapow replied:
I'm okay with looking at Apple's code in order to see how the API is meant to be used and such, I just don't want to subconsciously copy the code without realizing it. So, yes, I'll try to do this clean-room as much as I can.

Besides, the only real reason I'd have for checking Apple's code is stuff like this (i.e. internals that might affect external behavior, e.g. where I'm not sure if they allocate memory or not). For the most part, the APIs are pretty simple (once you consider RFCs/standards).

@LubosD Thanks for the info, I'll update it.

facekapow committed:
Use alloca instead of malloc

facekapow committed:
Fix typo in ccder_decode_bitstring

facekapow commented:
Upon further inspection, it seems that there's an issue with the `cczp_mod` function. Like I mentioned in darlinghq/darling#463, only `apple.com` passes the verification with the current implementation, and after generating some more test data, I've discovered that the issue is in `cczp_mod`.

I've been attacking this from different angles discovered that I had a bug in my code:

```c
while (ccn_cmp(extra_n, s, a) <= 0) {
```

Here, we're comparing `extra_n` units of `s` with `extra_n` units of `a`, but `a` only has `n` units. I honestly don't know how this wasn't causing a segfault, as `ccn_cmp` was accessing unallocated memory. Either way, once I fixed this by creating a copy of `a` with `extra_n` units and comparing `s` with the copy, the method once again slowed to a crawl, no better than a simple subtraction-loop-based modulo algorithm.

Considering what was discussed before, I decided to take a peek at Apple's CoreCrypto implementation and I think a comment finally revealed what the missing piece of the puzzle was:

```
Do r = a % d, where recip is the precalculated steady-state reciprocal of d
r is count cc_units in size, a is 2 * count units, d is count units
in size and recip is count + 1 units in size.
```

`cczp_mod` expects a *full* ZP structure, which includes a constant that I'm assuming (as I haven't tested it yet) is `floor((b^2k)/d)` where `b` is the base of the number (`2^64` in this case), `k` is the number of digits in the modulus (i.e. `n` in CoreCrypto), and `d` is the modulus. With this knowledge, I know that what I can do is try to implement Barrett reduction in `cczp_mod`. I'm currently working on this. Let's see how it goes...

facekapow committed:
Implement "_multi" variants of shift functions...

...also fix up the implementation of `ccn_div_euclid`

facekapow committed:
Properly implement `cczp_init` and `cczp_mod`

`cczp_init` is supposed to calculate the reciprocal of the prime its given and embed it into the ZP structure; `cczp_mod` now uses Barrett reduction to correctly calculate the remainder

facekapow committed:
Use the ZP structure from the key directly in `ccrsa_verify_pkcs1v15`

facekapow commented:
@darlinghq Is there a way to build Darling for debug but skip dSYM generation for some targets (e.g. JavaScriptCore)? Because having to generate dSYMs for *all* the modified targets on each build is slowing down my development (in fact, I've created a separate standalone copy of the library to build outside of Darling in order to quickly test iterations of my code as a temporary workaround).

CuriousTommy commented:
> Is there a way to build Darling for debug but skip dSYM generation for some targets (e.g. JavaScriptCore)?

I don't know about the dSYM part, but I believe that you can use the `-DFULL_BUILD=OFF` to avoid building the less important stuff.

facekapow committed:
Implement ccn_random_bits and ccn_make_recip...

...also fix the implementation of ccn_add to properly check for overflow

facekapow committed:
Use ccn_make_recip in cczp_init

facekapow committed:
Move reverse array printing functions to cc_debug.h

facekapow committed:
Initial elliptic curve cryptography functions

facekapow committed:
Implement ccecdh_compute_shared_secret...

...also implement ccrng and fix the implementations of ccn_write_int_size and ccn_write_uint_size

facekapow committed:
Implement ccec_export_pub and remove some debug statements

Those debug statements were printing out *way* too much information that we don't really need right now (because we haven't found any issues yet in the functions they debug)

facekapow committed:
Implement AES and AES-GCM

facekapow committed:
Fix SHA-384 and SHA-512

facekapow committed:
Remove debug messages

facekapow commented:
I've still got to clean up the implementation for most of this and try to optimize it, but I think this is finally out of WIP (but not 100% ready yet)! So far, with most websites I've tried, it works! (including GitHub, which is important for something I'll mention later on...).

BTW, if you're going to test it with cURL, be sure to compare results with Linux's cURL, since some websites don't return any data for cURL on *any* platform.

Like I said before, this is no longer WIP, but I've still got a few more things to add for some rarer cases. For example, I've still got to implement `ccec_verify`, necessary for some websites that use elliptic curve certificates (like https://www.wikipedia.org), but this is a rare case because the vast majority of the web use RSA certificates.

---

Also, the Homebrew installer now downloads and runs! (Because *of course* that was the first I was going to try 😄). Using the following command:
```bash
# slight modification of the command on brew.sh, because cURL thinks our certificate chain is "invalid"
/bin/bash -c "$(curl -fsSLk https://raw.githubusercontent.com/Homebrew/install/master/install.sh)"
```

...we get:
```
==> This script will install:
/usr/local/bin/brew
/usr/local/share/doc/homebrew
/usr/local/share/man/man1/brew.1
/usr/local/share/zsh/site-functions/_brew
/usr/local/etc/bash_completion.d/brew
/usr/local/Homebrew
==> The following new directories will be created:
/usr/local/etc
/usr/local/sbin
/usr/local/var
/usr/local/opt
/usr/local/share/zsh
/usr/local/share/zsh/site-functions
/usr/local/var/homebrew
/usr/local/var/homebrew/linked
/usr/local/Cellar
/usr/local/Caskroom
/usr/local/Homebrew
/usr/local/Frameworks

Press RETURN to continue or any other key to abort
Need sudo access on macOS!
This is a fake 'sudo' implementation, intended for use in Darling.
Processes will think they are run as UID/GID 0, but they are still run as your original UID/GID.
One purpose of this program is to convince some tools that they can write into '/'.
Another is to enable you to talk to certain system daemons.

Usage:
  sudo [-g GID] [-u UID] [-k] [--] COMMAND
```

So now the problem seems to be with the `sudo` command in Darling (which, after inspecting Homebrew's installation script, seems to be because `/usr/bin/sudo -l mkdir &>/dev/null` returns `1` instead of `0`).

---

And, after setting Git's `http.sslVerify` configuration option to `false` (for the same reason `-k` is needed for cURL) via `git config --global http.sslVerify false`, Git cloning for HTTPS URLs now works as well!

Cloning Darling within Darling :wink::
```
Darling [~/git]$ git clone https://www.github.com/darlinghq/darling
Cloning into 'darling'...
warning: redirecting to https://github.com/darlinghq/darling.git/
remote: Enumerating objects: 5, done.
remote: Counting objects: 100% (5/5), done.
remote: Compressing objects: 100% (5/5), done.
remote: Total 53002 (delta 0), reused 0 (delta 0), pack-reused 52997
Receiving objects: 100% (53002/53002), 27.56 MiB | 710.00 KiB/s, done.
Resolving deltas: 100% (34032/34032), done.
Checking out files: 100% (22217/22217), done.
```

Cloning the submodules recursively works as well, but just a heads up: Git cloning is pretty slow inside Darling (probably due to my unoptimized CoreCrypto code), so you're going to be there a while.

bugaevc commented:
Super cool! :tada: 

> cURL thinks our certificate chain is "invalid"

Any idea why?

> seems to be because `/usr/bin/sudo -l mkdir &>/dev/null` returns 1 instead of 0

Sounds like we're missing `-l`

> facekapow changed the title ~[WIP] Implement functions necessary for TLS/SSL~ Implement functions necessary for TLS/SSL

Note that this is still a "draft PR"; is that intentional?

facekapow commented:
> > cURL thinks our certificate chain is "invalid"
> 
> Any idea why?

If I had to guess, I'd say that either our certificates are outdated (however, this is unlikely), or something related to time/date is missing in Darling and causes cURL (and any other program using SSL) to think that our certificates have expired (more likely). However, this is just a guess; I'm going to look into this later.

> Note that this is still a "draft PR"; is that intentional?

I suppose I could make this a normal PR, I was just going to wait until I got all the web working 😄 (i.e. after adding support for elliptic curve certificates).

facekapow committed:
Fix conflict with origin/master

facekapow committed:
Fix DER decoding

facekapow committed:
Implement ccec_verify

facekapow commented:
Alright, I've implemented `ccec_verify` and now it seems that all websites work! 

As for the "invalid certificate chain" issue, there's a couple of issues in the way.

(I'm not sure where to put this next part since it's a couple different issues; maybe (a) new issue(s) in the central Darling repository? Either way, here it is:)

The first issue is that `_os_activity_initiate` isn't implemented. libSecurity calls `os_activity_initiate` (which is a macro wrapper for `_os_activity_initiate`) when it's going to communicate with securityd to verify the certificate chain. Since that function is a stub and does nothing, the `_trustResult` stays as `kSecTrustResultInvalid`.

So, in my local copy of Darling, I've implemented this method pretty simply and almost certainly not the way it's supposed to work (but it lets us progress further):
```
void _os_activity_initiate(void *dso, const char *description, os_activity_flag_t flags, os_block_t activity_block OS_NOESCAPE) {
	activity_block();
};
```

Should I make this a PR in libtrace?

The next issue comes from `_CFXPCCreateXPCObjectFromCFObject` not being implemented (and just returning `NULL`), causing the process to segfault. I'm currently working on implementing this and I'll submit two PRs (to CoreFoundation and libxpc) when I get it working (but I'm only going to implement the types needed for the securityd message; I'm (probably) not going to implement all the possible CF -> XPC conversions). The libxpc PR is necessary because some functions (specifically: `xpc_string_create`, `xpc_object_destroy`, `xpc_dictionary_set_value`, and `xpc_dictionary_destroy`) weren't working as expected: `xpc_string_create` should copy its input into a malloc-ed buffer, `xpc_object_destroy` needs to free the buffer in XPC strings, `xpc_dictionary_set_value` needs to copy the key given to it, and `xpc_dictionary_destroy` needs to free the keys in the dictionary.

There's probably something else in the way, but I'm going to take this one issue at a time 😄.

bugaevc commented:
> Alright, I've implemented `ccec_verify` and now it seems that all websites work!

🎉🎉

> Should I make this a PR in libtrace?

Yeah, please just add a comment that it's a stub.

About XPC: our libxpc (which is NextBSD-derived) needs a massive rewrite to make it more like Apple's — internally (it has to use libdispatch'es object model), and in the binary format. I was planning to do this myself, but if you want to work on that, please do. But my point is, don't waste too much time improving the existing libxpc since it's going to be rewritten soon anyway.

facekapow commented:
Do we have TrustD in Darling? Because if not, then I think I've determined the source of the "invalid certificate chain" issue. libSecurity tries to contact `com.apple.trustd` via `vproc_mig_look_up2` (indirectly), but it receives a `mig_reply_error_t` with a `RetCode` of `1102`, which is `BOOTSTRAP_UNKNOWN_SERVICE`. From what I can see, it doesn't look like Darling has it, which would explain the error.

Edit: On further investigation, it seems that `vproc_mig_look_up2` can't even lookup `com.apple.securityd`, which I'm *sure* is running. Weird.

Edit 2: `xtrace curl https://www.example.com`. Among other things:
```
[155] stat64("/Library/Preferences/com.apple.security.plist", 0x7fffffdfd5f8) -> ENOENT
```

bugaevc commented:
> On further investigation, it seems that vproc_mig_look_up2 can't even lookup com.apple.securityd, which I'm sure is running. Weird.

Really? I'm 100% sure curl used to talk to securityd successfully.

facekapow commented:
Well, xtrace and LLDB say:
```
[155] mach_msg_trap(0x7fffffdfd558, MACH_SEND_MSG|MACH_RCV_MSG, 64, 64, port 1543, 0, port 0)
[155]         {remote = copy send 1799, local = make send-once 1543, complex, id = 420}, ool [0x0; 0], 20 bytes of inline data
[155]         job::swap_complex(copy send 1799, 0, 6, mem [0x0; 0], 0)
[155]     mach_msg_trap() -> KERN_SUCCESS
[155]         {local = move send-once 1543, complex, id = 520}, ool [0x7ffff32ca000; 20971520], 12 bytes of inline data
[155]         job::swap_complex() -> mem [0x7ffff32ca000; 20971520], 20971520
[155] _kernelrpc_mach_vm_deallocate_trap(...)[155] task_self_trap() -> port right 259
[155] munmap(0x7ffff32ca000, 20971520) -> 0
 -> 0x
[155] _kernelrpc_mach_port_allocate_trap(task 259, MACH_PORT_RIGHT_RECEIVE) -> port right 4099
[155] _kernelrpc_mach_port_insert_right_trap(task 259, new name 4099, port right 4099, MACH_MSG_TYPE_MAKE_SEND) -> KERN_SUCCESS
[155] mach_msg_trap(0x7fffffdfd228, MACH_SEND_MSG|MACH_RCV_MSG, 188, 108, port 1543, 0, port 0)
[155]         {remote = copy send 1799, local = make send-once 1543, id = 404}, 164 bytes of inline data
[155]         job::look_up2(copy send 1799, "com.apple.securityd", 0, [138, 7, 0, 0, 153, 255, 255, 0, 0, 255, 0, 0, 0, 0, 0, 0], 0)
[155]     mach_msg_trap() -> KERN_SUCCESS
[155]         {local = move send-once 1543, id = 504}, 12 bytes of inline data
[155]         job::look_up2() -> BOOTSTRAP_UNKNOWN_SERVICE
```

LLDB in `vproc_mig_look_up2`
```
(lldb) p *(mig_reply_error_t *)Out0P)
(mig_reply_error_t) $0 = {
  Head = {
    msgh_bits = 4608
    msgh_size = 36
    msgh_remote_port = 0
    msgh_local_port = 1799
    msgh_voucher_port = 4162013280
    msgh_id = 504
  }
  NDR = {
    mig_vers = '\0'
    if_vers = '\0'
    reserved1 = '\0'
    mig_encoding = '\0'
    int_rep = '\x01'
    char_rep = '\0'
    float_rep = '\0'
    reserved2 = '\0'
  }
  RetCode = 1102
}
  Fix-it applied, fixed expression was:
    *(mig_reply_error_t *)Out0P
(lldb) p servicename
(char *) $1 = 0x00007ffff74f688a "com.apple.securityd"
```
(1102 == `BOOTSTRAP_UNKNOWN_SERVICE`)

facekapow committed:
Implement `ccder_decode_rsa_pub_x509_n`

facekapow committed:
Implement `ccrsa_import_pub`

facekapow committed:
Implement `ccrsa_pub_crypt`

facekapow committed:
Fix `ccn_write_uint_size`

Apple's version doesn't take into account the extra always-zero bit required to make the integer
always appear to be unsigned, so we can't take it into account either

facekapow committed:
Fix `ccrsa_import_pub`

We were accidentally using the start of the modulus body as the start of the exponent structure

facekapow committed:
Implement `ccrsa_emsa_pkcs1v15_verify`

facekapow committed:
Fix ccrsa_pub_crypt

We might have more units allocated in our key than in our input. This makes sure that we're only operating on the number of units actually used in the key

facekapow committed:
Fix ccn_write_uint

Write out the bytes starting at the least significant bytes. The way we were doing it before (starting at most significant bytes)
was leaving out the least significant bytes in cases were the most significant bytes were zero and users only provided enough space
for the bytes that were actually being used.

facekapow committed:
Implement `ccec_x963_import_pub_size`, `ccec_x963_import_priv_size`, `ccec_keysize_is_supported`, and `ccec_get_cp`

facekapow committed:
Fix DER size encoding and decoding

facekapow commented:
So after a *slew* of changes (to CoreFoundation, libxpc, syscall emulation, libtrace, and libSecurity), cURL finally works without the `-k` flag:
```
Darling [~]$ curl https://www.example.com
<!doctype html>
<html>
<head>
    <title>Example Domain</title>

    <!-- snipped for brevity -->
</head>

<body>
<div>
    <h1>Example Domain</h1>
    <p>This domain is for use in illustrative examples in documents. You may use this
    domain in literature without prior coordination or asking for permission.</p>
    <p><a href="https://www.iana.org/domains/example">More information...</a></p>
</div>
</body>
</html>
```
```
Darling [~]$ curl https://github.com






<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
  <link rel="dns-prefetch" href="https://github.githubassets.com">
  <link rel="dns-prefetch" href="https://avatars0.githubusercontent.com">
  <link rel="dns-prefetch" href="https://avatars1.githubusercontent.com">
  <link rel="dns-prefetch" href="https://avatars2.githubusercontent.com">
  <link rel="dns-prefetch" href="https://avatars3.githubusercontent.com">
  <link rel="dns-prefetch" href="https://github-cloud.s3.amazonaws.com">
  <link rel="dns-prefetch" href="https://user-images.githubusercontent.com/">



  <link crossorigin="anonymous" media="all" integrity="sha512-FG+rXqMOivrAjdEQE7tO4BwM1poGmg70hJFTlNSxjX87grtrZ6UnPR8NkzwUHlQEGviu9Xu
RYeO8zH9YwvZhdg==" rel="stylesheet" href="https://github.githubassets.com/assets/frameworks-146fab5ea30e8afac08dd11013bb4ee0.css" />
  <link crossorigin="anonymous" media="all" integrity="sha512-amiAO3vdUFnb2paOaFZ1w4h92dAdk9RCVFG5OhOY7iZfg4DSmhtO/EQ0BFRAeK9TCfUzbF8
7404O9UsVEi/K6g==" rel="stylesheet" href="https://github.githubassets.com/assets/site-6a68803b7bdd5059dbda968e685675c3.css" />
    <link crossorigin="anonymous" media="all" integrity="sha512-+zQhjOv4/Tzh5G/pTpd/Dqa3rv0YABLEui1mNHWdT/mnzvUmCWBx7P3V5Px/EMlo80v/7
a85JkV3fWw4zzotnA==" rel="stylesheet" href="https://github.githubassets.com/assets/github-fb34218cebf8fd3ce1e46fe94e977f0e.css" />






  <meta name="viewport" content="width=device-width">
<!-- the rest has been snipped for brevity -->
```
Including the report that GitHub is working is important for reasons outlined in darlinghq/darling-security#3 (basically, it's using a different type of certificate authority certificate).

## Trying this out yourself

The code necessary to get this all working (on top of the latest master I had it working on, darlinghq/darling@cd07ae854f09a89cd5f55234942c975a521e00a7; will test later with the most up to date master) is in:
  * this PR
  * darlinghq/darling-libtrace#1
  * darlinghq/darling-libxpc#6
  * darlinghq/darling-libxpc#7
  * darlinghq/darling-security#3

In addition, you need to copy `/System/Library/Security/Certificates.bundle` from a real macOS installation to your Darling prefix (usually `~/.darling`, so the full path would be `~/.darling/System/Library/Security/Certificates.bundle`).

## Homebrew

Yep. 😉

The Homebrew install no longer fails on `sudo -l` (thanks, @ahyattdev!). In a fresh Darling prefix, with only the `Certificates.bundle` copied in:
```
Darling [~]$ /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)"
==> This script will install:
/usr/local/bin/brew
/usr/local/share/doc/homebrew
/usr/local/share/man/man1/brew.1
/usr/local/share/zsh/site-functions/_brew
/usr/local/etc/bash_completion.d/brew
/usr/local/Homebrew
==> The following new directories will be created:
/usr/local/bin
/usr/local/etc
/usr/local/include
/usr/local/lib
/usr/local/sbin
/usr/local/var
/usr/local/opt
/usr/local/share/zsh
/usr/local/share/zsh/site-functions
/usr/local/var/homebrew
/usr/local/var/homebrew/linked
/usr/local/Cellar
/usr/local/Caskroom
/usr/local/Homebrew
/usr/local/Frameworks
==> The Xcode Command Line Tools will be installed.

Press RETURN to continue or any other key to abort
==> /usr/bin/sudo /bin/mkdir -p /usr/local/bin /usr/local/etc /usr/local/include /usr/local/lib /usr/local/sbin /usr/local/var /usr/local/opt /usr/local/share/zsh /usr/local/share/zsh/site-functions /usr/local/var/homebrew /usr/local/var/homebrew/linked /usr/local/Cellar /usr/local/Caskroom /usr/local/Homebrew /usr/local/Frameworks
==> /usr/bin/sudo /bin/chmod g+rwx /usr/local/bin /usr/local/etc /usr/local/include /usr/local/lib /usr/local/sbin /usr/local/var /usr/local/opt /usr/local/share/zsh /usr/local/share/zsh/site-functions /usr/local/var/homebrew /usr/local/var/homebrew/linked /usr/local/Cellar /usr/local/Caskroom /usr/local/Homebrew /usr/local/Frameworks
==> /usr/bin/sudo /usr/sbin/chown ariel /usr/local/bin /usr/local/etc /usr/local/include /usr/local/lib /usr/local/sbin /usr/local/va
r /usr/local/opt /usr/local/share/zsh /usr/local/share/zsh/site-functions /usr/local/var/homebrew /usr/local/var/homebrew/linked /usr
/local/Cellar /usr/local/Caskroom /usr/local/Homebrew /usr/local/Frameworks
chown: ariel: illegal user name
Failed during: /usr/bin/sudo /usr/sbin/chown ariel /usr/local/bin /usr/local/etc /usr/local/include /usr/local/lib /usr/local/sbin /u
sr/local/var /usr/local/opt /usr/local/share/zsh /usr/local/share/zsh/site-functions /usr/local/var/homebrew /usr/local/var/homebrew/
linked /usr/local/Cellar /usr/local/Caskroom /usr/local/Homebrew /usr/local/Frameworks
```
Welp, looks like Darling's `chown` can't handle the current user's username. Weird, but it should be pretty simple enough to patch.

Anyways, I decided to try running it again in the same prefix. This time the install progressed a *lot* further (and succeeded? I think?):
```
Darling [~]$ /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)"
==> This script will install:
/usr/local/bin/brew
/usr/local/share/doc/homebrew
/usr/local/share/man/man1/brew.1
/usr/local/share/zsh/site-functions/_brew
/usr/local/etc/bash_completion.d/brew
/usr/local/Homebrew
==> The following existing directories will be made writable by user only:
/usr/local/share/zsh
/usr/local/share/zsh/site-functions
==> The Xcode Command Line Tools will be installed.

Press RETURN to continue or any other key to abort
==> /usr/bin/sudo /bin/chmod u+rwx /usr/local/share/zsh /usr/local/share/zsh/site-functions
==> /usr/bin/sudo /bin/chmod 755 /usr/local/share/zsh /usr/local/share/zsh/site-functions
==> /usr/bin/sudo /bin/mkdir -p /Users/ariel/Library/Caches/Homebrew
==> Searching online for the Command Line Tools
==> /usr/bin/sudo /usr/bin/touch /tmp/.com.apple.dt.CommandLineTools.installondemand.in-progress
==> Installing the Command Line Tools (expect a GUI popup):
==> /usr/bin/sudo /usr/bin/xcode-select --install
You are about to download and install Apple Command Line Tools covered by the following license:
https://www.apple.com/legal/sla/docs/xcode.pdf

Do you agree with the terms of the license? (y/n) y
Downloading packages...
Downloading CLTools_Executables.pkg...
Installing...
installer: Installing package com.apple.pkg.CLTools_Executables version 9.2.0.0.1.1510905681 (597942 KB)
installer: Extracting files
installer: Installation complete
Downloading CLTools_macOS_SDK_Remove_1012.pkg...
Installing...
installer: Installing package com.apple.pkg.CLTools_SDK_OSX1012 version 9.2.0.0.1.1510905681 (94 KB)
installer: Extracting files
installer: Installation complete
Downloading CLTools_SDK_macOS1013.pkg...
Installing...
installer: Installing package com.apple.pkg.CLTools_SDK_macOSSDK version 9.2.0.0.1.1510905681 (313954 KB)
installer: Extracting files
installer: Installation complete
Downloading CLTools_macOS_SDK_Remove_1013.pkg...
Installing...
installer: Installing package com.apple.pkg.CLTools_SDK_macOS1013 version 9.2.0.0.1.1510905681 (94 KB)
installer: Extracting files
installer: Installation complete
Downloading DevSDK_OSX1012.pkg...
Installing...
installer: Installing package com.apple.pkg.DevSDK_OSX1012 version 9.2.0.0.1.1510905681 (261534 KB)
installer: Extracting files
installer: Installation complete
Installation complete!
Press any key when the installation has completed.
==> /usr/bin/sudo /usr/bin/xcode-select --switch /Library/Developer/CommandLineTools
==> Downloading and installing Homebrew...
remote: Enumerating objects: 43, done.
remote: Counting objects: 100% (43/43), done.
remote: Compressing objects: 100% (43/43), done.
remote: Total 134716 (delta 0), reused 43 (delta 0), pack-reused 134673
Receiving objects: 100% (134716/134716), 32.42 MiB | 727.00 KiB/s, done.
Resolving deltas: 100% (98972/98972), done.
From https://github.com/Homebrew/brew
 * [new branch]          master     -> origin/master
 * [new tag]             0.1        -> 0.1
 * [new tag]             0.2        -> 0.2
 * [new tag]             0.3        -> 0.3
 * [new tag]             0.4        -> 0.4
 * [new tag]             0.5        -> 0.5
 * [new tag]             0.6        -> 0.6
 * [new tag]             0.7        -> 0.7
 * [new tag]             0.7.1      -> 0.7.1
 * [new tag]             0.8        -> 0.8
 * [new tag]             0.8.1      -> 0.8.1
 * [new tag]             0.9        -> 0.9
 * [new tag]             0.9.1      -> 0.9.1
 * [new tag]             0.9.2      -> 0.9.2
 * [new tag]             0.9.3      -> 0.9.3
 * [new tag]             0.9.4      -> 0.9.4
 * [new tag]             0.9.5      -> 0.9.5
 * [new tag]             0.9.8      -> 0.9.8
 * [new tag]             0.9.9      -> 0.9.9
 * [new tag]             1.0.0      -> 1.0.0
 * [new tag]             1.0.1      -> 1.0.1
 * [new tag]             1.0.2      -> 1.0.2
 * [new tag]             1.0.3      -> 1.0.3
 * [new tag]             1.0.4      -> 1.0.4
 * [new tag]             1.0.5      -> 1.0.5
 * [new tag]             1.0.6      -> 1.0.6
 * [new tag]             1.0.7      -> 1.0.7
 * [new tag]             1.0.8      -> 1.0.8
 * [new tag]             1.0.9      -> 1.0.9
 * [new tag]             1.1.0      -> 1.1.0
 * [new tag]             1.1.1      -> 1.1.1
 * [new tag]             1.1.10     -> 1.1.10
 * [new tag]             1.1.11     -> 1.1.11
 * [new tag]             1.1.12     -> 1.1.12
 * [new tag]             1.1.13     -> 1.1.13
 * [new tag]             1.1.2      -> 1.1.2
 * [new tag]             1.1.3      -> 1.1.3
 * [new tag]             1.1.4      -> 1.1.4
 * [new tag]             1.1.5      -> 1.1.5
 * [new tag]             1.1.6      -> 1.1.6
 * [new tag]             1.1.7      -> 1.1.7
 * [new tag]             1.1.8      -> 1.1.8
 * [new tag]             1.1.9      -> 1.1.9
 * [new tag]             1.2.0      -> 1.2.0
 * [new tag]             1.2.1      -> 1.2.1
 * [new tag]             1.2.2      -> 1.2.2
 * [new tag]             1.2.3      -> 1.2.3
 * [new tag]             1.2.4      -> 1.2.4
 * [new tag]             1.2.5      -> 1.2.5
 * [new tag]             1.2.6      -> 1.2.6
 * [new tag]             1.3.0      -> 1.3.0
 * [new tag]             1.3.1      -> 1.3.1
 * [new tag]             1.3.2      -> 1.3.2
 * [new tag]             1.3.3      -> 1.3.3
 * [new tag]             1.3.4      -> 1.3.4
 * [new tag]             1.3.5      -> 1.3.5
 * [new tag]             1.3.6      -> 1.3.6
 * [new tag]             1.3.7      -> 1.3.7
 * [new tag]             1.3.8      -> 1.3.8
 * [new tag]             1.3.9      -> 1.3.9
 * [new tag]             1.4.0      -> 1.4.0
 * [new tag]             1.4.1      -> 1.4.1
 * [new tag]             1.4.2      -> 1.4.2
 * [new tag]             1.4.3      -> 1.4.3
 * [new tag]             1.5.0      -> 1.5.0
 * [new tag]             1.5.1      -> 1.5.1
 * [new tag]             1.5.10     -> 1.5.10
 * [new tag]             1.5.11     -> 1.5.11
 * [new tag]             1.5.12     -> 1.5.12
 * [new tag]             1.5.13     -> 1.5.13
 * [new tag]             1.5.14     -> 1.5.14
 * [new tag]             1.5.2      -> 1.5.2
 * [new tag]             1.5.3      -> 1.5.3
 * [new tag]             1.5.4      -> 1.5.4
 * [new tag]             1.5.5      -> 1.5.5
 * [new tag]             1.5.6      -> 1.5.6
 * [new tag]             1.5.7      -> 1.5.7
 * [new tag]             1.5.8      -> 1.5.8
 * [new tag]             1.5.9      -> 1.5.9
 * [new tag]             1.6.0      -> 1.6.0
 * [new tag]             1.6.1      -> 1.6.1
 * [new tag]             1.6.10     -> 1.6.10
 * [new tag]             1.6.11     -> 1.6.11
 * [new tag]             1.6.12     -> 1.6.12
 * [new tag]             1.6.13     -> 1.6.13
 * [new tag]             1.6.14     -> 1.6.14
 * [new tag]             1.6.15     -> 1.6.15
 * [new tag]             1.6.16     -> 1.6.16
 * [new tag]             1.6.17     -> 1.6.17
 * [new tag]             1.6.2      -> 1.6.2
 * [new tag]             1.6.3      -> 1.6.3
 * [new tag]             1.6.4      -> 1.6.4
 * [new tag]             1.6.5      -> 1.6.5
 * [new tag]             1.6.6      -> 1.6.6
 * [new tag]             1.6.7      -> 1.6.7
 * [new tag]             1.6.8      -> 1.6.8
 * [new tag]             1.6.9      -> 1.6.9
 * [new tag]             1.7.0      -> 1.7.0
 * [new tag]             1.7.1      -> 1.7.1
 * [new tag]             1.7.2      -> 1.7.2
 * [new tag]             1.7.3      -> 1.7.3
 * [new tag]             1.7.4      -> 1.7.4
 * [new tag]             1.7.5      -> 1.7.5
 * [new tag]             1.7.6      -> 1.7.6
 * [new tag]             1.7.7      -> 1.7.7
 * [new tag]             1.8.0      -> 1.8.0
 * [new tag]             1.8.1      -> 1.8.1
 * [new tag]             1.8.2      -> 1.8.2
 * [new tag]             1.8.3      -> 1.8.3
 * [new tag]             1.8.4      -> 1.8.4
 * [new tag]             1.8.5      -> 1.8.5
 * [new tag]             1.8.6      -> 1.8.6
 * [new tag]             1.9.0      -> 1.9.0
 * [new tag]             1.9.1      -> 1.9.1
 * [new tag]             1.9.2      -> 1.9.2
 * [new tag]             1.9.3      -> 1.9.3
 * [new tag]             2.0.0      -> 2.0.0
 * [new tag]             2.0.1      -> 2.0.1
 * [new tag]             2.0.2      -> 2.0.2
 * [new tag]             2.0.3      -> 2.0.3
 * [new tag]             2.0.4      -> 2.0.4
 * [new tag]             2.0.5      -> 2.0.5
 * [new tag]             2.0.6      -> 2.0.6
 * [new tag]             2.1.0      -> 2.1.0
 * [new tag]             2.1.1      -> 2.1.1
 * [new tag]             2.1.10     -> 2.1.10
 * [new tag]             2.1.11     -> 2.1.11
 * [new tag]             2.1.12     -> 2.1.12
 * [new tag]             2.1.13     -> 2.1.13
 * [new tag]             2.1.14     -> 2.1.14
 * [new tag]             2.1.15     -> 2.1.15
 * [new tag]             2.1.16     -> 2.1.16
 * [new tag]             2.1.2      -> 2.1.2
 * [new tag]             2.1.3      -> 2.1.3
 * [new tag]             2.1.4      -> 2.1.4
 * [new tag]             2.1.5      -> 2.1.5
 * [new tag]             2.1.6      -> 2.1.6
 * [new tag]             2.1.7      -> 2.1.7
 * [new tag]             2.1.8      -> 2.1.8
 * [new tag]             2.1.9      -> 2.1.9
 * [new tag]             2.2.0      -> 2.2.0
 * [new tag]             2.2.1      -> 2.2.1
 * [new tag]             2.2.10     -> 2.2.10
 * [new tag]             2.2.11     -> 2.2.11
 * [new tag]             2.2.12     -> 2.2.12
 * [new tag]             2.2.13     -> 2.2.13
 * [new tag]             2.2.2      -> 2.2.2
 * [new tag]             2.2.3      -> 2.2.3
 * [new tag]             2.2.4      -> 2.2.4
 * [new tag]             2.2.5      -> 2.2.5
 * [new tag]             2.2.6      -> 2.2.6
 * [new tag]             2.2.7      -> 2.2.7
 * [new tag]             2.2.8      -> 2.2.8
 * [new tag]             2.2.9      -> 2.2.9
HEAD is now at e8d626556 Merge pull request #7380 from dawidd6/fix-docker-tag
/System/Library/Frameworks/Ruby.framework/Versions/Current/usr/bin/ruby: warning: unknown argument for --enable: `frozen-string-literal'
/System/Library/Frameworks/Ruby.framework/Versions/Current/usr/bin/ruby: warning: unknown argument for --disable: `did_you_mean'
-e:1:in `<main>': undefined method `canonical_segments' for #<Gem::Version "2.0.0"> (NoMethodError)
==> Downloading https://homebrew.bintray.com/bottles-portable-ruby/portable-ruby-2.6.3.mavericks.bottle.tar.gz
######################################################################## 100.0%
Error: Cannot verify the checksum ('shasum' or 'sha256sum' not found)!
Error: Failed to install vendor Ruby.
Failed during: /usr/local/bin/brew update --force
==> Installation successful!

==> Homebrew has enabled anonymous aggregate formulae and cask analytics.
Read the analytics documentation (and how to opt-out) here:
  https://docs.brew.sh/Analytics
No analytics data has been sent yet (or will be during this `install` run).

==> Homebrew is run entirely by unpaid volunteers. Please consider donating:
  https://github.com/Homebrew/brew#donations

==> Next steps:
- Run `brew help` to get started
- Further documentation:
    https://docs.brew.sh
```
Unfortunately, though the install *says* it succeeds, it does not. Trying to run `brew install wget`:
```
Darling [~]$ brew install wget
/System/Library/Frameworks/Ruby.framework/Versions/Current/usr/bin/ruby: warning: unknown argument for --enable: `frozen-string-literal'
/System/Library/Frameworks/Ruby.framework/Versions/Current/usr/bin/ruby: warning: unknown argument for --disable: `did_you_mean'
-e:1:in `<main>': undefined method `canonical_segments' for #<Gem::Version "2.0.0"> (NoMethodError)
==> Downloading https://homebrew.bintray.com/bottles-portable-ruby/portable-ruby-2.6.3.mavericks.bottle.tar.gz
Already downloaded: /Users/ariel/Library/Caches/Homebrew/portable-ruby-2.6.3.mavericks.bottle.tar.gz
Error: Cannot verify the checksum ('shasum' or 'sha256sum' not found)!
Error: Failed to install vendor Ruby.
```
After a bit of googling, it looks like Homebrew expects there to be a `shasum` command, which we could probably emulate with a script that uses `openssl sha`.

## Up Next

Having to copy over the certificate bundle from a macOS installation is No Bueno. I'm currently working on a way to generate this bundle ourselves as a part of the Darling build.

The cryptography part (i.e. this PR) is also slow as molasses right now, so I've got to optimize that.

bugaevc commented:
👍👍👍

facekapow commented:
Whoops, little extra note about the Homebrew section in the comment above: `chown` only seems to be broken when run with `sudo`:
```
Darling [~]$ mkdir tmp
Darling [~]$ chown ariel tmp
Darling [~]$ sudo chown ariel tmp
chown: ariel: illegal user name
Darling [~]$
```
